File: src/main/java/dev/sky_lock/pocketlifevehicle/vehicle/CarArmorStand.java

package dev.sky_lock.pocketlifevehicle.vehicle;

import dev.sky_lock.pocketlifevehicle.task.BurnExplosionTask;
import dev.sky_lock.pocketlifevehicle.task.SubmergedMessageTask;
import net.minecraft.server.v1_14_R1.*;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.craftbukkit.v1_14_R1.CraftServer;
import org.bukkit.craftbukkit.v1_14_R1.entity.CraftArmorStand;
import org.bukkit.craftbukkit.v1_14_R1.entity.CraftEntity;
import org.bukkit.craftbukkit.v1_14_R1.entity.CraftPlayer;
import org.bukkit.entity.ArmorStand;
import org.bukkit.util.EulerAngle;

/**
 * @author sky_lock
 */

public class CarArmorStand extends EntityArmorStand {

    private Car car;

    public CarArmorStand(EntityTypes<? extends EntityArmorStand> entitytypes, World world) {
        super(entitytypes, world);
    }

    public CarArmorStand(World world, double x, double y, double z) {
        super(world, x, y, z);
        NBTTagCompound nbt = new NBTTagCompound();
        nbt.setBoolean("NoBasePlate", true);
        nbt.setBoolean("Invulnerable", true);
        nbt.setBoolean("PersistenceRequired", true);
        nbt.setBoolean("NoGravity", false);
        nbt.setBoolean("Invisible", true);
        nbt.setBoolean("Marker", false); // ArmorStand has a very small collision box when true
        nbt.setBoolean("Small", true);
        this.a(nbt);
        //乗れるブロックの高さ
        this.K = 1.0F;
    }

    void assemble(Car car) {
        this.car = car;
        ArmorStand armorstand = (CraftArmorStand) this.getBukkitEntity();
        armorstand.setRightArmPose(EulerAngle.ZERO);
        armorstand.setItem(car.getModel().getItemOption().getPosition().getSlot(), car.getModel().getItemStack());
        // armorstand.setMetadata("mocar-as", new FixedMetadataValue(PLVehicle.getInstance(), null));
        car.getSound().start();
    }

    //TODO: 当たり判定
//    @Override
//    public EntitySize a(EntityPose entitypose) {
//        EntitySize size = this.getEntityType().j();
//        CollideBox collideBox = this.car.getModel().getCollideBox();
//        float widthScale = collideBox.getBaseSide() / size.width;
//        float heightScale = collideBox.getHeight() / size.height;
//        return this.getEntityType().j().a(widthScale, heightScale);
//    }

    public Location getLocation() {
        return getBukkitEntity().getLocation();
    }

    @Override
    public void killEntity() {
        super.killEntity();
        car.getSound().stop();
    }

    //ツタ、はしご、足場ブロックを登れなくする
    @Override
    public boolean isClimbing() {
        return false;
    }

    //足音がなるかどうか
    @Override
    public boolean isSilent() {
        return isCarArmorStand();
    }

    //水に入った時
    @Override
    protected void az() {
        if (!isCarArmorStand()) {
            super.au();
            return;
        }
        super.au();
        new SubmergedMessageTask().run(car);
        car.getSound().stop();
    }

    @Override
    protected void burn(float i) {
        if (!isCarArmorStand()) {
            super.burn(i);
            return;
        }
        super.burn(i);
        if (!car.isBeginExplode()) {
            new BurnExplosionTask().run(car);
            car.setBeginExplode(true);
        }
    }

    @Override
    public void e(Vec3D vec3d) {
        if (car.getPassengers().isEmpty() || !car.getDriver().isPresent() || this.isInWater() || this.inLava) {
            car.getEngine().stop();
            super.e(vec3d);
            return;
        }

        car.getDriver().ifPresent(driver -> {
            EntityPlayer player = ((CraftPlayer) driver).getHandle();
            float sideIn = player.bb;
            float forwardIn = player.bd;

            if (sideIn < 0.0F) {
                car.getSteering().right(driver);
            } else if (sideIn > 0.0F) {
                car.getSteering().left(driver);
            }

            car.getEngine().update(forwardIn);
            car.getEngine().consumeFuel(sideIn);
        });

        this.fallDistance = 0.0F;

        this.yaw = car.getStatus().getYaw();
        this.lastYaw = this.yaw;
        this.pitch = 0.0F;
        setYawPitch(this.yaw, this.pitch);
        // this.aQ = this.yaw;

        this.o(car.getEngine().getCurrentSpeed());
        super.e(vec3d.e(new Vec3D(0.0, 1.0, 3.0)));
        car.getStatus().setLocation(getLocation());
    }

    @Override
    public CraftEntity getBukkitEntity() {
        return new CraftCar((CraftServer) Bukkit.getServer(), this);
    }

    private boolean isCarArmorStand() {
        return car.getModel() != null && car.getStatus() != null;
    }

    public class CraftCar extends CraftArmorStand {
        CraftCar(CraftServer server, EntityArmorStand entity) {
            super(server, entity);
        }

        public CarArmorStand getHandle() {
            return (CarArmorStand) super.getHandle();
        }
    }

}

File: src/main/kotlin/dev/sky_lock/pocketlifevehicle/vehicle/CarArmorStand.kt
package dev.sky_lock.pocketlifevehicle.vehicle

import dev.sky_lock.pocketlifevehicle.task.BurnExplosionTask
import dev.sky_lock.pocketlifevehicle.task.SubmergedMessageTask
import net.minecraft.server.v1_14_R1.*
import org.bukkit.Location
import org.bukkit.craftbukkit.v1_14_R1.entity.CraftArmorStand
import org.bukkit.craftbukkit.v1_14_R1.entity.CraftPlayer
import org.bukkit.entity.ArmorStand
import org.bukkit.entity.Player
import org.bukkit.util.EulerAngle

/**
 * @author sky_lock
 */
class CarArmorStand : EntityArmorStand {
    private var car: Car? = null

    constructor(entitytypes: EntityTypes<out EntityArmorStand>, world: World) : super(entitytypes, world) {}
    constructor(world: World, x: Double, y: Double, z: Double) : super(world, x, y, z) {
        val nbt = NBTTagCompound()
        nbt.setBoolean("NoBasePlate", true)
        nbt.setBoolean("Invulnerable", true)
        nbt.setBoolean("PersistenceRequired", true)
        nbt.setBoolean("NoGravity", false)
        nbt.setBoolean("Invisible", true)
        nbt.setBoolean("Marker", false) // ArmorStand has a very small collision box when true
        nbt.setBoolean("Small", true)
        this.a(nbt)
        //乗れるブロックの高さ
        K = 1.0f
    }

    fun assemble(car: Car) {
        this.car = car
        val armorstand: ArmorStand = bukkitEntity as CraftArmorStand
        armorstand.rightArmPose = EulerAngle.ZERO
        val model = car.model
        armorstand.setItem(model.itemOption.position.slot, model.itemStack)
        armorstand.isSmall = !model.isBig
        // armorstand.setMetadata("mocar-as", new FixedMetadataValue(PLVehicle.getInstance(), null));
        car.soundTask.start()
    }

    //TODO: 当たり判定
    //    @Override
    //    public EntitySize a(EntityPose entitypose) {
    //        EntitySize size = this.getEntityType().j();
    //        CollideBox collideBox = this.car.getModel().getCollideBox();
    //        float widthScale = collideBox.getBaseSide() / size.width;
    //        float heightScale = collideBox.getHeight() / size.height;
    //        return this.getEntityType().j().a(widthScale, heightScale);
    //    }
    val location: Location
        get() = bukkitEntity.location

    override fun killEntity() {
        car!!.soundTask.stop()
        super.killEntity()
    }

    //ツタ、はしご、足場ブロックを登れなくする
    override fun isClimbing(): Boolean {
        return false
    }

    //足音がなるかどうか
    override fun isSilent(): Boolean {
        return isCarArmorStand
    }

    //水に入った時
    override fun az() {
        if (!isCarArmorStand) {
            super.au()
            return
        }
        super.au()
        SubmergedMessageTask().run(car!!)
        car!!.soundTask.stop()
    }

    override fun burn(i: Float) {
        if (!isCarArmorStand) {
            super.burn(i)
            return
        }
        super.burn(i)
        if (!car!!.isBeginExplode) {
            BurnExplosionTask().run(car!!)
            car!!.isBeginExplode = true
        }
    }

    override fun e(vec3d: Vec3D) {
        if (car!!.passengers.isEmpty() || !car!!.driver.isPresent || this.isInWater || inLava) {
            car!!.engine.stop()
            super.e(vec3d)
            return
        }
        car!!.driver.ifPresent { driver: Player ->
            val player = (driver as CraftPlayer).handle
            val sideIn = player.bb
            val forwardIn = player.bd
            if (sideIn < 0.0f) {
                car!!.steering.right(driver)
            } else if (sideIn > 0.0f) {
                car!!.steering.left(driver)
            }
            car!!.engine.update(forwardIn)
            car!!.engine.consumeFuel(sideIn)
        }
        fallDistance = 0.0f
        yaw = car!!.status.yaw
        lastYaw = yaw
        pitch = 0.0f
        setYawPitch(yaw, pitch)
        // this.aQ = this.yaw;
        this.o(car!!.engine.currentSpeed)
        super.e(vec3d.e(Vec3D(0.0, 1.0, 3.0)))
        car!!.status.location = location
    }

//    override fun getBukkitEntity(): CraftEntity {
//        return CraftCar(Bukkit.getServer() as CraftServer, this)
//    }

    private val isCarArmorStand: Boolean
        get() = car!!.model != null && car!!.status != null

/*    inner class CraftCar internal constructor(server: CraftServer, entity: EntityArmorStand) : CraftArmorStand(server, entity) {
        override fun getHandle(): CarArmorStand {
            return super.getHandle() as CarArmorStand
        }
    }*/
}